### PR TITLE
EVG-12450: Allow FTDC to collect JSON in memory

### PR DESCRIPTION
**Pull request by maravichandran**

### CODE REVIEW

File: metrics/json.go [hunk 1]
```diff
@@ -126,58 +126,65 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
+			return output, nil
+		} else {
+			fn := fmt.Sprintf("%s.%d", opts.OutputFilePrefix, outputCount)
+			if err = ioutil.WriteFile(fn, output, 0600); err != nil {
+				return nil, errors.Wrapf(err, "problem writing data to file %s", fn)
+			}
 		}
 
 		outputCount++
 		collector.Reset()
```
Comment (Kimchelly) [inline metrics/json.go:162]: It seems like these statements (`collector.Reset()` and `flushTimer.Reset(opts.FlushInterval)` should still run even if we return early (e.g. when `opts.OutputFilePrefix == ""`). Can you put these in a defer?

File: metrics/json.go [hunk 2]
```diff
@@ -126,58 +126,65 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
 
-		return nil
+		return nil, nil
 	}
 
 	docs, errs := opts.getSource()
 
 	for {
 		select {
+		//case <-
```
Comment (Kimchelly) [inline metrics/json.go:172]: Remove comment.

File: metrics/json.go [hunk 3]
```diff
@@ -126,58 +126,65 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
 		flushTimer.Reset(opts.FlushInterval)
 
-		return nil
+		return nil, nil
 	}
 
 	docs, errs := opts.getSource()
 
 	for {
 		select {
```
Comment (Kimchelly) [inline metrics/json.go:171]: It seems like all of the statements in this select block will always return a nil result. In which cases is it supposed to return a non-nil byte slice?

File: metrics/json.go [hunk 4]
```diff
@@ -126,58 +126,65 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
+			return nil, errors.WithStack(err)
 		case doc := <-docs:
 			if err := collector.Add(doc); err != nil {
-				return errors.Wrap(err, "problem collecting results")
+				return nil, errors.Wrap(err, "problem collecting results")
 			}
 		case <-flushTimer.C:
-			return errors.Wrap(flusher(), "problem flushing results at the end of the file")
+			_, errFlusher := flusher()
+			return nil, errors.Wrap(errFlusher, "problem flushing results at the end of the file")
```
Comment (Kimchelly) [inline metrics/json.go:187]: The error wrap should probably be "flushing results" without implying that it's writing to a file.

File: metrics/json_test.go [hunk 5]
```diff
@@ -187,15 +187,32 @@ func TestCollectJSON(t *testing.T) {
...
 
+		require.NoError(t, writeStream(hundredDocs, f))
+		require.NoError(t, f.Close())
+
+		opts := CollectJSONOptions{
+			FileName:    fn,
+			SampleCount: 100,
+		}
+
+		_, err = CollectJSONStream(ctx, opts)
```
Comment (Kimchelly) [inline metrics/json_test.go:207]: This should verify that the resulting byte stream is what you expect it to be.

File: metrics/json_test.go [hunk 6]
```diff
@@ -214,7 +231,7 @@ func TestCollectJSON(t *testing.T) {
 			Follow:        true,
 		}
 
-		err = CollectJSONStream(ctx, opts)
+		_, err = CollectJSONStream(ctx, opts)
```
Comment (Kimchelly) [inline metrics/json_test.go:234]: Check the return value.

File: metrics/json_test.go [hunk 7]
```diff
@@ -260,7 +277,7 @@ func TestCollectJSON(t *testing.T) {
 		}
 		ctx := context.Background()
 
-		err = CollectJSONStream(ctx, opts)
+		_, err = CollectJSONStream(ctx, opts)
```
Comment (Kimchelly) [inline metrics/json_test.go:280]: Check the return value.

File: metrics/json.go (hunk 2 above)
Comment (maravichandran) [reply metrics/json.go:172]: done

File: metrics/json.go [hunk 8]
```diff
@@ -126,58 +126,65 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
 
-	flusher := func() error {
-		fn := fmt.Sprintf("%s.%d", opts.OutputFilePrefix, outputCount)
+	flusher := func() ([]byte, error) {
 		info := collector.Info()
 
 		if info.SampleCount == 0 {
 			flushTimer.Reset(opts.FlushInterval)
-			return nil
+			return nil, nil
```
[the remark this answers is not in the document]
Comment (maravichandran) [reply metrics/json.go:144]: done

File: metrics/json.go (hunk 4 above)
Comment (maravichandran) [reply metrics/json.go:187]: done

File: metrics/json_test.go (hunk 6 above)
Comment (maravichandran) [reply metrics/json_test.go:234]: done

File: metrics/json_test.go (hunk 7 above)
Comment (maravichandran) [reply metrics/json_test.go:280]: done

File: metrics/json.go (hunk 1 above)
Comment (maravichandran) [reply metrics/json.go:162]: done

File: metrics/json.go (hunk 3 above)
Comment (maravichandran) [reply metrics/json.go:173]: done

File: metrics/json_test.go (hunk 5 above)
Comment (maravichandran) [reply metrics/json_test.go]: done

File: metrics/json.go [hunk 9]
```diff
@@ -126,58 +126,66 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
 
 	for {
 		select {
 		case <-ctx.Done():
-			return errors.New("operation aborted")
+			return nil, errors.New("operation aborted")
 		case err := <-errs:
 			if err == nil || errors.Cause(err) == io.EOF {
-				return errors.Wrap(flusher(), "problem flushing results at the end of the file")
+				output, errFlusher := flusher()
```
Comment (Kimchelly) [inline metrics/json.go:178]: Can we just call this `err`?

File: metrics/json.go [hunk 10]
```diff
@@ -126,58 +126,66 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
-			return errors.WithStack(err)
+			return nil, errors.WithStack(err)
 		case doc := <-docs:
 			if err := collector.Add(doc); err != nil {
-				return errors.Wrap(err, "problem collecting results")
+				return nil, errors.Wrap(err, "problem collecting results")
 			}
 		case <-flushTimer.C:
-			return errors.Wrap(flusher(), "problem flushing results at the end of the file")
+			output, errFlusher := flusher()
```
Comment (Kimchelly) [inline metrics/json.go:187]: Same thing about error name.

File: metrics/json.go [hunk 11]
```diff
@@ -126,58 +126,66 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
+			fn := fmt.Sprintf("%s.%d", opts.OutputFilePrefix, outputCount)
+			if err = ioutil.WriteFile(fn, output, 0600); err != nil {
+				return nil, errors.Wrapf(err, "problem writing data to file %s", fn)
+			}
 		}
 
 		outputCount++
-		collector.Reset()
-		flushTimer.Reset(opts.FlushInterval)
+		defer func() {
```
Comment (Kimchelly) [inline metrics/json.go:162]: The flush timer reset has to go at the beginning of the function. `collector.Reset()` has to be done after the collector is resolved.

File: metrics/json.go (hunk 9 above)
Comment (maravichandran) [reply metrics/json.go:178]: done

File: metrics/json.go (hunk 10 above)
Comment (maravichandran) [reply metrics/json.go:187]: done

File: metrics/json.go (hunk 11 above)
Comment (maravichandran) [reply metrics/json.go:164]: done

File: metrics/json.go [hunk 12]
```diff
@@ -126,58 +126,68 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
+
```
Comment (Kimchelly) [inline metrics/json.go:165]: Remove defer.

File: metrics/json.go [hunk 13]
```diff
@@ -126,58 +126,68 @@ func (opts CollectJSONOptions) getSource() (<-chan *birch.Document, <-chan error
...
 	if err := opts.validate(); err != nil {
-		return errors.WithStack(err)
+		return nil, errors.WithStack(err)
 	}
 
 	outputCount := 0
 	collector := ftdc.NewDynamicCollector(opts.SampleCount)
 	flushTimer := time.NewTimer(opts.FlushInterval)
 	defer flushTimer.Stop()
+	defer flushTimer.Reset(opts.FlushInterval)
```
Comment (Kimchelly) [inline metrics/json.go:138]: This should defer within `flusher`.

File: metrics/json.go (hunk 12 above)
Comment (maravichandran) [reply metrics/json.go:165]: done

File: metrics/json.go (hunk 13 above)
Comment (maravichandran) [reply metrics/json.go]: done

**Pull request merged**